core/light-client/src/concurrent_access.rs

```diff
@@ -0,0 +1,145 @@
+use std::sync::RwLock;
+
+use crate::{
+	error::{Error, Result},
+	LightClientState, Validator,
+};
+use finality_grandpa::BlockNumberOps;
+use itp_sgx_io::SealedIO;
+use lazy_static::lazy_static;
+use sp_runtime::traits::{Block as BlockT, NumberFor};
+use std::marker::PhantomData;
+
+lazy_static! {
+	// As long as we have a file backend, we use this 'dummy' lock,
+	// which guards against concurrent read/write access.
+	pub static ref VALIDATOR_LOCK: RwLock<()> = Default::default();
+}
+
+/// Retrieve an exclusive lock on a validator for either read or write access.
+///
+/// In order to hide the whole locks mechanics, we provide an interface that allows executing
+/// either a mutating, or a non-mutating function on the validator.
+/// The reason we have this additional wrapper around `SealedIO`, is that we need
+/// to guard against concurrent access by using RWLocks (which `SealedIO` does not do).
+pub trait ValidatorAccess<PB>
+where
+	PB: BlockT,
+	NumberFor<PB>: BlockNumberOps,
+{
+	type ValidatorType: Validator<PB> + LightClientState<PB>;
+
+	fn execute_on_validator<F, R>(&self, getter_function: F) -> Result<R>
+	where
+		F: FnOnce(&Self::ValidatorType) -> Result<R>;
+
+	fn execute_mut_on_validator<F, R>(&self, mutating_function: F) -> Result<R>
+	where
+		F: FnOnce(&mut Self::ValidatorType) -> Result<R>;
+}
+
+/// Implementation of a validator access based on a global lock and corresponding file.
+#[derive(Clone, Debug)]
+pub struct GlobalValidatorAccessor<ValidatorT, PB, Seal>
+where
+	ValidatorT: Validator<PB> + LightClientState<PB>,
+	Seal: SealedIO<Error = Error, Unsealed = ValidatorT>,
+	PB: BlockT,
+	NumberFor<PB>: BlockNumberOps,
+{
+	_phantom: PhantomData<(Seal, ValidatorT, PB)>,
+}
+
+impl<ValidatorT, PB, Seal> Default for GlobalValidatorAccessor<ValidatorT, PB, Seal>
+where
+	ValidatorT: Validator<PB> + LightClientState<PB>,
+	Seal: SealedIO<Error = Error, Unsealed = ValidatorT>,
+	PB: BlockT,
+	NumberFor<PB>: BlockNumberOps,
+{
+	fn default() -> Self {
+		GlobalValidatorAccessor { _phantom: Default::default() }
+	}
+}
+
+impl<ValidatorT, PB, Seal> GlobalValidatorAccessor<ValidatorT, PB, Seal>
+where
+	ValidatorT: Validator<PB> + LightClientState<PB>,
+	Seal: SealedIO<Error = Error, Unsealed = ValidatorT>,
+	PB: BlockT,
+	NumberFor<PB>: BlockNumberOps,
+{
+	pub fn new() -> Self {
+		GlobalValidatorAccessor { _phantom: Default::default() }
+	}
+}
+
+impl<ValidatorT, PB, Seal> ValidatorAccess<PB> for GlobalValidatorAccessor<ValidatorT, PB, Seal>
+where
+	ValidatorT: Validator<PB> + LightClientState<PB>,
+	Seal: SealedIO<Error = Error, Unsealed = ValidatorT>,
+	PB: BlockT,
+	NumberFor<PB>: BlockNumberOps,
+{
+	type ValidatorType = ValidatorT;
+
+	fn execute_on_validator<F, R>(&self, getter_function: F) -> Result<R>
+	where
+		F: FnOnce(&Self::ValidatorType) -> Result<R>,
+	{
+		let _read_lock = VALIDATOR_LOCK.read().map_err(|_| Error::PoisonedLock)?;
+		let validator = Seal::unseal()?;
+		getter_function(&validator)
```

Comment:
The non-mutating version, basically just a getter, uses only `unseal` with a read lock (of which there can be multiple concurrent).